packages/eslint-plugin/src/rules/no-unsafe-unary-minus.ts

```diff
@@ -0,0 +1,55 @@
+import type * as ts from 'typescript';
+
+import * as util from '../util';
+
+interface TypeChecker extends ts.TypeChecker {
+  // https://github.com/microsoft/TypeScript/issues/9879
+  isTypeAssignableTo(source: ts.Type, target: ts.Type): boolean;
+  getUnionType(types: ts.Type[]): ts.Type;
+}
+
+type Options = [];
+type MessageIds = 'unaryMinus';
+
+export default util.createRule<Options, MessageIds>({
+  name: 'no-unsafe-unary-minus',
+  meta: {
+    type: 'problem',
+    docs: {
+      description: 'Require unary negation to take a number',
+      requiresTypeChecking: true,
+    },
+    messages: {
+      unaryMinus: 'Invalid type "{{type}}" of template literal expression.',
+    },
+    schema: [],
+  },
+  defaultOptions: [],
+  create(context) {
+    return {
+      UnaryExpression(node): void {
+        if (node.operator !== '-') {
+          return;
+        }
+        const services = util.getParserServices(context);
+        const argType = services.getTypeAtLocation(node.argument);
+        const checker = services.program.getTypeChecker() as TypeChecker;
+        if (
+          !checker.isTypeAssignableTo(
+            argType,
+            checker.getUnionType([
+              checker.getNumberType(), // first exposed in TypeScript v5.1
+              checker.getBigIntType(), // first added in TypeScript v5.1
+            ]),
```

Comment:
...that being said, I don't think you need the type assignability APIs for this specific rule. I think you should be able to just compare with `ts.TypeFlags`. Search on examples of `isTypeFlagSet` and `ts.TypeFlags.Num` in code for examples.

Reply:
Does this actually work? I wrote an initial version using just `TypeFlags`, and it failed even on a few of the simple test cases I currently have. Are you saying `TypeFlags.Num` is just straight-up equivalent to being assignable to `number | bigint`, or is there more logic I'd need to write on top of that?

Reply:
I haven't tried it locally, but in the past I've had to use variants like `TypeFlags.NumberLike` (see `no-unsafe-enun-comparison`) and/or `TypeFlags.BigIntLike` (see `restrict-template-expressions`).

Reply:
Ah OK; I tried those in my initial implementation and it didn't work on unions, for instance. (And I believe `any` and literals, but those have their own flags.) I didn't go further down that route because I don't have any idea how more complicated types like intersections or mapped types etc would interact with it.

Reply:
> unions

Ah yes unions can be tricky. Rules often have to `tsutils.unionTypeParts` to get them, then run a check on each of the constituents.

> intersections or mapped types

Fortunately, you probably don't need to do anything special for them. `getTypeAtLocation` will retrieve whatever the result of their type is.

---

And, this reminds me - there should also be some tests for generics / type parameters. `util.getConstrainedTypeAtLocation` should be used.

* `<T>(t: T) => -t` -> that might be unsafe, since we don't know what `T` is
* `<T extends number>(t: T) => -t` -> that isn't unsafe, since although we don't know what `T` is, we do know it's _constrained_ to `number`